CRM: Create Automations UI root

robertf4 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->

Fixes https://github.com/Automattic/zero-bs-crm/issues/3179

This PR adds the root of a React app for Automations and creates a new Automations module to load and serve it.

<img width="1171" alt="image" src="https://github.com/Automattic/jetpack/assets/42627630/7cd440a3-d7aa-40c0-a51a-782cbd28f8d9">

**It aim to do so by:**

* Register "Automations" as a core module
* Add feature flag to prevent the UI to be exposed.
* Finish the CRM specific `AdminPage` component
  * We also correct an older dummy test for the OBW component that was used to test if React builds worked and was tested.
* Add a new `<AutomationAdmin>` component that will be the main component for the automations view.

**Caveats:**

* The RNA container component only supports the narrow Jetpack view - not the ~1551px one that was recently used in Stats and CRM. I will create a separate PR to address this issue in the RNA package.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
p1HpG7-mCz-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1. Add the following to a mu-plugin (e.g. `0-sandbox.php`): `add_filter( 'jetpack_crm_feature_flag_automations', '__return_true' );`
2. Run `pnpm build` in the CRM folder.
3. Visit `/wp-admin/admin.php?page=jpcrm-automations` and verify that you see a page with an "Automations" header like the screenshot above.
4. In the CRM folder run `pnpm run test` and verify that all Automations tests pass.
5. Verify that Automations shows up in the menu bar.
6. Disable the feature flag filter and verify that Automations does not show up in the menu bar and that the page is not accessible.

robertf4 committed:
[not verified] Initial admin root

robertf4 committed:
[not verified] changelog

robertf4 committed:
[not verified] Add header

robertf4 committed:
[not verified] Add feature flag

robertf4 committed:
[not verified] Add tests

robertf4 commented on projects/plugins/crm/modules/automations/jpcrm-automations-init.php:
```diff
@@ -0,0 +1,118 @@
+
+	add_submenu_page(
+		$zbs->slugs['automations'],
+		__( 'Welcome to Jetpack CRM Automations', 'zero-bs-crm' ), // TODO: change this
```
I put the TODO here as I'm not sure what this copy should be, so this is just a reasonable placeholder.

kallehauge replied:
The existing routes just say e.g. `Quotes`.
Maybe it's about time we update it to prefix it with e.g.: `Jetpack CRM: `, so in this case it would just be `Jetpack CRM: Automations`.

The main objective here is to make it easy for folks to navigate their browser tabs and doesn't have to apply to any SEO related rules

robertf4 replied:
Once I got the menu to work I found that this title is what is used in the menu:

<img width="163" alt="image" src="https://github.com/Automattic/jetpack/assets/42627630/2125a3d4-b72a-4da5-b208-7ffaa71cf757">

So I figured it is best as Automations.

robertf4 committed:
[not verified] Update TODO

kallehauge commented on projects/plugins/crm/src/js/components/automations/styles.scss:
```diff
@@ -0,0 +1,9 @@
+
+.jpcrm-automations-admin__subheader {
+    color: $studio-gray-60;
+}
```
Missing new line 🆕 📏

kallehauge commented on projects/plugins/crm/modules/automations/jpcrm-automations-init.php:
```diff
@@ -0,0 +1,118 @@
+<?php
```
**Request**: I assume that this was converted from a class to function due to "the magical logic that loads the init file", but would it be cleaner if we separated this into two files?

1. `class-admin-page.php`: Contains an `Admin_Page` class with 90% of the code from this file.  
2. `jpcrm-automations-init.php` Basically just calls `Admin_Page->init()` together with the feature flag (there might be some magical loading logic I'm forgetting).

By doing the above, we can also init other classes when we e.g. start to bootstrap the automations engine and it doesn't have to be mixed together with admin page related code

robertf4 replied:
I made it functions instead of a class because there is no state here. So in the onboarding wizard there is a singleton pattern and accessing the singleton causes all of the init logic to run, but it is all functions with no state so I decided to just make this functions.

Until there is state I'd prefer to keep things as functions just so the semantics of the code match the functionailty, but I can certainly begin splitting out this logic into more files.

robertf4 replied:
I split out all the admin page init logic into `admin-page-init.php`.

kallehauge commented on projects/plugins/crm/modules/automations/jpcrm-automations-init.php:
```diff
@@ -0,0 +1,118 @@
+function register_page() {
+	global $zbs;
+
+	add_submenu_page(
+		$zbs->slugs['automations'],
+		__( 'Welcome to Jetpack CRM Automations', 'zero-bs-crm' ), // TODO: verify or get better copy
+		'',
+		'manage_options',
+		'jpcrm-automations',
+		'Automattic\Jetpack_CRM\Automations\render_page'
+	);
```
**Observation**: The `Jetpack CRM` menu item isn't "active" while we're on this page. 

I also believe that `$zbs->slugs['automations']` can/should be used instead of the hardcoded value.

```suggestion
	add_submenu_page(
		$zbs->slugs['home'],
		__( 'Welcome to Jetpack CRM Automations', 'zero-bs-crm' ), // TODO: verify or get better copy
		'',
		'manage_options',
		$zbs->slugs['automations'],
		'Automattic\Jetpack_CRM\Automations\render_page'
	);
```

The above example might not work, and if it doesn't, then it's 💯 to "just" register a task on the React navigation router issue so we can continue to debug it there

![Screenshot 2023-07-13 at 22 36 24](https://github.com/Automattic/jetpack/assets/3846700/d3be5d8d-9384-4f18-b4b6-9f7f4936c486)

robertf4 replied:
In investigating this I found that CRM has a whole complex system for building the menu in `ZeroBSCRM.Core.Menus.WP.php`. I was able to get the menu to work by integrating the Automations page into this system.

To accomplish this, the rendering of the page had to be moved to `admin/automations/main.page.php` while the rest of the init logic is in `admin-page-init.php` due to the order that certain hooks are called.

Initially I didn't like the separation, but this pattern does make more sense in the overall context of CRM code than what I was doing. It also makes sense for the logic in `admin-page-init.php` to remain in the module as it is responsible for adding the initial state which will require using business logic that is likely to be shared with other code in the module.

kallehauge commented on projects/plugins/crm/src/js/components/automations/index.tsx:
```diff
@@ -0,0 +1,14 @@
+
+import './styles.scss';
+
+export const AutomationsAdmin = () => {
```
The file name and component name doesn't match.

Referencing it as "Admin" or "Page" is otherwise 👍 with me

robertf4 replied:
I switched the file name.

kallehauge commented on projects/plugins/crm/src/js/components/automations/index.tsx:
```diff
@@ -0,0 +1,14 @@
+
+export const AutomationsAdmin = () => {
+	return (
+		<div className="jpcrm-automations-admin__container">
```
Should we perhaps reuse the `<AdminPage>` RNA component since it makes sure we show a footer etc? 😄

If we use the `AdminPage` component that already exist in CRM (which wraps the RNA component) we should be able to only define e.g. the footer in 1 location.

**References**:

* `projects/plugins/crm/src/js/components/admin-page/index.tsx`
* `projects/js-packages/components/components/admin-page/index.tsx`

robertf4 replied:
In this case we already have the header and footer:

<img width="1168" alt="image" src="https://github.com/Automattic/jetpack/assets/42627630/e7719ecf-4a9f-41a1-b37e-d73f808a8228">

Those are both existing functionality written in PHP and added to the page by CRM, so if we switch to the RNA component then we would have to recreate it in React and also change CRM to not render this header and footer for the automations page. This could be something worth doing, especially if we want to start making more pages in React, but I think it's a project on its own.

robertf4 replied:
For reference, this is what it looks like when I use `AdminPage` without making any other changes to CRM. The headers and footers are doubled.

<img width="1167" alt="image" src="https://github.com/Automattic/jetpack/assets/42627630/f7794c6d-9376-4398-8e2e-56a100ad2184">

kallehauge replied:
The admin component accepts `showHeader` and `showFooter` props to hide them. I've gone ahead and made a small refactor so it's reusable for future admin pages as well.

kallehauge commented on projects/plugins/crm/src/js/components/automations/view.tsx:
```diff
@@ -0,0 +1,23 @@
+		return;
+	}
+
+	const component = <AutomationsAdmin />;
```
AFAIK then `ThemeProvider` should make sure that we have support/access to all Jetpack colors (e.g.: `--jp-black-80`)

```suggestion
	// @todo: Remove fallback when we drop support for WP 6.1
	const component = (
		<ThemeProvider>
			<AutomationsAdmin />
		</ThemeProvider>
	);
```

robertf4 replied:
I'm not exactly sure what `ThemeProvider` is giving us as I still need to have `@import '../../../../sass/_color_studio.scss';` in the scss file in order for it to compile, but I added it anyway.

kallehauge replied:
`<ThemeProvider>` makes sure that the `<AutomationsAdmin>` component can use the CSS variables Jetpack exposes:

![Screenshot 2023-07-18 at 15 41 28](https://github.com/Automattic/jetpack/assets/3846700/06e735d8-582d-430e-aa3b-d5a368aff7ee)

robertf4 committed:
[not verified] Add newline

robertf4 committed:
[not verified] Fix menu by using Core.Menus

robertf4 committed:
[not verified] Revert extraneous changes

robertf4 committed:
Rename automations component to automations-admin

robertf4 committed:
Create admin-page-init.php

robertf4 committed:
Use ThemeProvider

robertf4 committed:
Update changelog

robertf4 committed:
Add feature flags to menus

kallehauge committed:
Adhere to monorepo expectations

kallehauge committed:
Automation: move admin init file into admin dir

kallehauge committed:
Prevent direct file access to automations main page

kallehauge committed:
Correct load admin page hook

kallehauge committed:
Rename core automations slug

We currently have a paid extension using the same slug, so if that has been enabled, we're having issues with overrides.
E.g.: the CRM system will not recognize the new page as an admin page.

kallehauge committed:
Refactor admin page to use a CRM generic admin page wrapper

kallehauge committed:
Remove unused style file

kallehauge committed:
Merge branch 'trunk' into add/crm/automations-admin-root
